xgcm/padding.py

```diff
@@ -0,0 +1,306 @@
+                        # after all this logic, start slicing
+                        # we get the appropriate slice to add and
+                        # remove the previously padded values from the target
+                        source_slice = source_da.isel({source_dim: source_slice_index})
+
+                        target_slice = da_single_padded.isel(
+                            {target_dim: target_slice_index}
+                        )
+
+                        # rename dimension if different
+                        if source_dim != target_dim:
+                            if not reverse:
+                                # Flip along the orthogonal axis
+                                source_slice = source_slice.isel(
+                                    {target_dim: slice(None, None, -1)}
+                                )
+                            source_slice = _maybe_swap_dimension_names(
+                                source_slice,
+                                source_dim,
+                                target_dim,
+                            )
+
+                        # Apply parallel flip if reverse
+                        if reverse:
+                            source_slice = source_slice.isel(
+                                {target_dim: slice(None, None, -1)}
+                            )
+                        source_slice = source_slice.squeeze()
+
+                        # assemble the padded array
+                        if is_right:
+                            concat_list = [target_slice, source_slice]
+                        else:
+                            concat_list = [source_slice, target_slice]
+
+                        da_single_padded = xr.concat(
+                            concat_list,
+                            dim=target_dim,
+                            coords="minimal",
+                        )
+                        # TODO: Can we do this with an assignment in xarray? Maybe not important yet.
+        da_split.append(da_single_padded)
+
+    da_padded = xr.concat(da_split, dim=facedim)
+
+    # trim back to original shape
+    def _trim_expanded_padding_width(da, grid, padding_width, padding_width_expanded):
+        for axname in padding_width.keys():
+            _, dim = grid.axes[axname]._get_position_name(da)
+            start = padding_width_expanded[axname][0] - padding_width[axname][0]
+            stop = padding_width_expanded[axname][1] - padding_width[axname][1]
+            # if stop is zero we want to index the full end of the dimension
+            if stop == 0:
+                stop = None
+            else:
+                stop = -stop
+            da = da.isel({dim: slice(start, stop)})
+        return da
+
+    return _trim_expanded_padding_width(
+        da_padded, grid, padding_width_original, padding_width
+    )
+
+
+def _pad_basic(da, grid, padding_width, padding="fill", fill_value=None):
+    """Implement basic xarray/numpy padding methods"""
```

Comment:
Ideally we would eventually remove this, otherwise we're going to end up with two different code paths that can do the same thing.

Reply:
Also we it would be a good idea to look at how `xarray.pad`/`dask.pad` works internally in case we can improve dask performance somehow.